# SPDX-License-Identifier: FSL-1.1-ALv2
# rules_oci 2.3.0: a source script and its generated Windows launcher are not
# ancestor/descendant paths. Resolve between files, give different scripts
# distinct launchers, and create the shared descriptor launcher only once.
diff --git a/oci/private/util.bzl b/oci/private/util.bzl
--- a/oci/private/util.bzl
+++ b/oci/private/util.bzl
@@ -3 +3 @@
-load("@bazel_skylib//lib:paths.bzl", "paths")
+load("@aspect_bazel_lib//lib:paths.bzl", "relative_file")
@@ -208 +208 @@
-    win_launcher = ctx.actions.declare_file("wrap_%s.bat" % ctx.label.name)
+    win_launcher = ctx.actions.declare_file("wrap_%s_%s.bat" % (ctx.label.name, bash_launcher.basename))
@@ -227 +227 @@
-            launcher = paths.relativize(bash_launcher.path, win_launcher.dirname),
+            launcher = relative_file(bash_launcher.path, win_launcher.path),
diff --git a/oci/private/image.bzl b/oci/private/image.bzl
--- a/oci/private/image.bzl
+++ b/oci/private/image.bzl
@@ -107 +107 @@
-def _calculate_descriptor(ctx, idx, layer, zstd, jq, coreutils, regctl):
+def _calculate_descriptor(ctx, idx, layer, zstd, jq, coreutils, regctl, launcher):
@@ -113,4 +113,4 @@
     ctx.actions.run(
-        executable = util.maybe_wrap_launcher_for_windows(ctx, ctx.executable._descriptor_sh),
-        inputs = [layer],
+        executable = launcher,
+        inputs = [layer, ctx.executable._descriptor_sh],
         outputs = [descriptor],
@@ -191,4 +191,5 @@
     # add layers
+    descriptor_launcher = util.maybe_wrap_launcher_for_windows(ctx, ctx.executable._descriptor_sh)
     for (i, layer) in enumerate(ctx.files.tars):
-        descriptor = _calculate_descriptor(ctx, i, layer, zstd, jq, coreutils, regctl)
+        descriptor = _calculate_descriptor(ctx, i, layer, zstd, jq, coreutils, regctl, descriptor_launcher)
         inputs.append(descriptor)
